embedded-cal-nrf54l15/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+                in_desc.sz = sz(4);
+                in_desc.dmatag = 19;
+
+                state_desc.addr = state.as_ptr() as *mut u8;
+                state_desc.next = &mut block_desc;
+                state_desc.sz = sz(32);
+                state_desc.dmatag = 99;
+
+                block_desc.addr = instance.block.as_ptr() as *mut u8;
+                block_desc.next = &mut data_desc;
+                block_desc.sz = instance.block_bytes_used as u32;
+                block_desc.dmatag = 3;
+
+                data_desc.addr = data.as_ptr() as *mut u8;
+                data_desc.next = LAST_DESC_PTR;
+                data_desc.sz = 0x2000_0000 | bytes_from_data as u32;
+                data_desc.dmatag = 35;
+            }
+        }
+
+        dma.fetchaddrlsb()
+            .write(|w| unsafe { w.bits((&mut in_desc) as *mut _ as u32) });
+
+        dma.pushaddrlsb()
+            .write(|w| unsafe { w.bits((&mut out_desc) as *mut _ as u32) });
+
+        dma.config().write(|w| {
+            w.fetchctrlindirect().set_bit();
+            w.pushctrlindirect().set_bit();
+            w.fetchstop().clear_bit();
+            w.pushstop().clear_bit();
+            w.softrst().clear_bit()
+        });
+
+        dma.start().write(|w| {
+            w.startfetch().set_bit();
+            w.startpush().set_bit()
+        });
+
+        while dma.status().read().fetchbusy().bit_is_set() {}
+        while dma.status().read().pushbusy().bit_is_set() {}
+
+        instance.state = Some(new_state);
+        instance.processed_bytes += instance.block_bytes_used + bytes_from_data;
+
+        // reset buffer
+        instance.block = [0u8; BLOCK_SIZE];
+        instance.block_bytes_used = 0;
+
+        // copy leftover bytes into empty buffer
+        let data_left = data.len() - bytes_from_data;
+        instance.block[0..data_left].copy_from_slice(&data[bytes_from_data..]);
+        instance.block_bytes_used += data_left;
+    }
+
+    fn finalize(&mut self, instance: Self::HashState) -> Self::HashResult {
+        let dma = self.p.global_cracencore_s.cryptmstrdma();
+
+        let mut pad: [u8; 128] = [0x00; 128];
+        let padding_size = sha256_padding(
+            instance.processed_bytes + instance.block_bytes_used,
+            &mut pad,
+        );
+
+        let mut out: [u8; 32] = [0x00; 32];
+
+        let mut out_desc = Descriptor {
+            addr: out.as_mut_ptr(),
+            next: LAST_DESC_PTR,
+            sz: sz(32),
+            dmatag: 32,
+        };
+
+        let header: [u8; 4] = match instance.state {
+            Some(_) => [instance.algorithm as u8, 0x04, 0x00, 0x00],
+            None => [instance.algorithm as u8, 0x06, 0x00, 0x00],
+        };
+
+        let mut in_desc = Descriptor::empty();
+        let mut state_desc = Descriptor::empty();
+        let mut data_desc = Descriptor::empty();
+        let mut padding_desc = Descriptor::empty();
+
+        match instance.state {
```

Comment:
That looks familiar from above. Can't the padding data be created, sent into `self.update()`, with an assertion that (by construction of the padding) there is no data left in the buffer after it?

Reply:
The issue (that AIU still persists) is that there is a large amount of duplication between the update and the finish function. This would be way more readable and less prone to editing errors if some of that update/finish code were refactored into (non-trait) methods that are then used commonly between update and finalize.